### PR TITLE
Add support for MPS

### PR DESCRIPTION
**Pull request by Developer-Ecosystem-Engineering**
We've added support for OIDN_MPS on configurations where MPS is available. This results in at least a 10x performance improvement over CPU only

- Added odinTests for MPS
- Updated cmake file to detect and use MPS when available
- Added MPSGraphNetwork & MPS Devices

### Test Results

On a 2.6 GHz 6-Core Intel Core i7 with AMD Radeon Pro 5600M

- CPU
  - Benchmarking: ntimes=20 sec=45.4701, msec/image=2273.51

- MPS
  - Benchmarking: ntimes=20 sec=2.91587, msec/image=145.794

RT.hdr_alb_nrm.1920x1080 ... 1058.66 (145.004) msec/image
RT.ldr_alb_nrm.1920x1080 ... 1086.15 (144.358) msec/image
RT.hdr_alb_nrm.3840x2160 ... 4518.11 (614.316) msec/image
RT.ldr_alb_nrm.3840x2160 ... 4495.98 (608.582) msec/image
RT.hdr_alb_nrm.1280x720 ... 468.137 (66.82) msec/image
RT.ldr_alb_nrm.1280x720 ... 470.525 (65.8203) msec/image
RTLightmap.hdr.2048x2048 ... 2045.76 (234.115) msec/image
RTLightmap.hdr.4096x4096 ... 9731.88 (1021.19) msec/image
RTLightmap.hdr.1024x1024 ... 544.029 (58.3557) msec/image

### CODE REVIEW

**Comment by abebeos** [comment]
a "pixel-perfect" PR.

@atafra, why is this not even getting a comment?

**Comment by abebeos** [comment]
> a "pixel-perfect" PR.

@Developer-Ecosystem-Engineering , while the speedtests impress, having all changes within one commit makes the review difficult.

I would split this down into multiple commits.

File: .gitmodules [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [submodule "mkl-dnn"]
 	path = mkl-dnn
-	url = ../mkl-dnn.git
+url=https://github.com/OpenImageDenoise/mkl-dnn.gitOpenImageDenoise/mkl-dnn.git
```
Comment (abebeos) [inline .gitmodules:3]: You should not changes existent build behavior. Here you break the fork behavior.

File: CMakeLists.txt [hunk 2]
```diff
@@ -240,6 +267,8 @@ endif()
 generate_cpp_from_blob(WEIGHTS_SOURCES "oidn::blobs::weights" ${WEIGHTS})
 target_sources(${PROJECT_NAME} PRIVATE ${WEIGHTS_SOURCES})
 
+include_directories(BEFORE ${PROJECT_SOURCE_DIR}/core/)
+
```
Comment (abebeos) [inline CMakeLists.txt:271]: * Change of behavior.
* undocumented.

**Comment by atafra** [comment]
@abebeos This will be available in the upcoming Open Image Denoise 2.0.

File: apps/oidnTest.cpp [hunk 3]
```diff
@@ -57,35 +57,43 @@ TEST_CASE("single filter", "[single_filter]")
 {
   const int W = 257;
   const int H = 89;
+  
+  std::vector<DeviceType> deviceType = { DeviceType::CPU };
+#if defined(OIDN_MPS)
+  deviceType.push_back(DeviceType::MPS);
+#endif
+
```
Comment (abebeos) [inline apps/oidnTest.cpp:65]: Aborting review. Too many undocumented general changes (outside of a #if defined(OIDN_MPS))

**Review by abebeos — CHANGES_REQUESTED** [review]
Replay this pr

-less general changes
-more commits
-more comments

**Comment by abebeos** [comment]
> @abebeos This will be available in the upcoming Open Image Denoise 2.0.

Nice. Where can one follow the development? And do you have an estimated release date?

**Comment by atafra** [comment]
The development for 2.0 is currently internal. It is planned to be released in about a month.

**Comment by atafra** [comment]
Not yet but it will be released very soon now.

**Pull request closed without merging**